src/compiler/program.ts

```diff
@@ -160,15 +162,22 @@ namespace ts {
         host = host || createCompilerHost(options);
 
         let filesByName = createFileMap<SourceFile>(fileName => host.getCanonicalFileName(fileName));
-
-        forEach(rootNames, name => processRootFile(name, /*isDefaultLib:*/ false));
-
-        // Do not process the default library if:
-        //  - The '--noLib' flag is used.
-        //  - A 'no-default-lib' reference comment is encountered in
-        //      processing the root files.
-        if (!skipDefaultLib) {
-            processRootFile(host.getDefaultLibFileName(options), /*isDefaultLib:*/ true);
+        
+        // if old program was provided by has different target module kind - assume that it cannot be reused
+        // different module kind can lead to different way of resolving modules
+        if (oldProgram && oldProgram.getCompilerOptions().module !== options.module) {

```

Comment:
because in case if structure cannot be reused we still want to utilize previous resolutions, see `processImports`. However if old version of the program targeted different module kind - then resolutions should not be shared (not true currently but will be true after we incorporate different module resolution strategies)